AEP/aep-00004.md

```diff
@@ -0,0 +1,163 @@
+# AEP 4: Refactoring of MongoDB module : possibility to deploy a sharded database and use of Helm charts
+|                   |ArmoniK Enhancement Proposal|
+---:                |:---
+**AEP**             | 4
+**Title**           | AEP 4: Shard the MongoDB database and deploy it with a Helm chart
+**Author**          | Trystan Schneider <<tschneider@aneo.fr>>
+**Status**          | Draft
+**Type**            | Standard
+**Creation Date**   | 2024-06-26
+
+# Abstract
+This AEP proposes a new deployment for the MongoDB database. The insight is to shard (horizontal partitioning) MongoDB and deploy it with a Helm chart (the package manager for Kubernetes).
+
+# Motivation
+The motivation of this AEP is to proactively treat a previsible bottleneck on the database. 
+
+Currently in ArmoniK, MongoDB is deployed with a simple replica set architecture, meaning that the database is composed of multiple identical instances. However, only one instance of the database supports write operations (the primary node), resulting in a potential bottleneck on this node. Additionnaly replicating the whole database is not very efficient for read operations, as the DB engine has to parse bigger datasets.
+
+Finally, the current MongoDB module in our infrastructure's code is defined kind of manually with multiple resources being defined, which is not  necessarily the best choice as deploying a database, especially a sharded one, is a complex task. 
```

Comment:
I am not sure to understant what your are writing here

Reply:
Currently, the way the database is deployed entirely replicates data with multiple identical replicas. And among these replicas, only one supports write operations (inserting data into the database). This where the concern of a potential bottleneck lies. Also, I wanted to mention that a non-sharded database could raise a concern on read operations too.

For the second paragraph I was highlighting that our current MongoDB deployment, which doesn't use a Helm chart, is not using the best practices, as we handle the complexity of creating all the resources needed for a MongoDB server ourselves, instead of delegating it to a Helm chart.